src/finn/transformation/change_3d_tensors_to_4d.py

```diff
@@ -0,0 +1,219 @@
+        "Conv",
+        "Transpose",
+        "LogSoftmax",
+        "ArgMax",
+    ]
+    invalid_nodes = []
+    for n in model.graph.node:
+        node_op_type = n.op_type
+        if node_op_type in valid_nodes:
+            continue
+        else:
+            invalid_nodes.append(node_op_type)
+
+    return invalid_nodes
+
+
+class Change3DTo4DTensors(Transformation):
+    """
+    Replaces 3D tensors with 4D tensors assuming the following format:
+    [N, C, H] -> [N, C, H, 1].
+    The attributes of a (specific) set of supported nodes are changed accordingly.
+    If the graph contains unsupported nodes, a warning is raised and the transformation
+    is not applied.
+    """
+
+    def apply(self, model):
+        graph_modified = False
+
+        invalid_nodes = _find_invalid_nodes(model)
+        if len(invalid_nodes) > 0:
+            warnings.warn(
+                "Transformation is not applied,\
+                 found unsupported nodes in the graph: {}.".format(
+                    invalid_nodes
+                )
+            )
+            return (model, graph_modified)
+
+        # Infer the shapes of each tensor, remove unused tensors
+        # and give each tensor a readable name
+        model = model.transform(InferShapes())
+        model = model.transform(RemoveUnusedTensors())
+        model = model.transform(GiveUniqueNodeNames())
+        model = model.transform(GiveReadableTensorNames())
+
+        # Converts 3D tensors representing the input, a value_info, or the output
+        # to 4D tensors
+        # Inputs
+        tensor_names = {}
+        for t in model.graph.input:
+            tensor_name = t.name
+            tensor_type = t.type.tensor_type.elem_type
+            tensor_shape = model.get_tensor_shape(tensor_name)
+            tensor_names[tensor_name] = [tensor_type]
+            tensor_names[tensor_name].append(tensor_shape)
+
+        # Initializers
+        initializer_names = []  # list with all initializers
+        initializers = {}  # list with initializers from Conv, Mul, and Add nodes
+        for i in model.graph.initializer:
+            init_name = i.name
+            initializer_names.append(init_name)
+            if "Conv" in init_name:
+                init_dim = i.dims
+                init_dtype = i.data_type
+                initializers[init_name] = [init_dtype]
+                initializers[init_name].append(init_dim)
+            elif init_name[0:4] == "Mul_":
+                init_dim = i.dims
+                if len(i.dims) == 3:
+                    init_dtype = i.data_type
+                    initializers[init_name] = [init_dtype]
+                    initializers[init_name].append(init_dim)
+            elif "Add" in init_name:
+                init_dim = i.dims
+                if len(i.dims) == 3:
+                    init_dtype = i.data_type
+                    initializers[init_name] = [init_dtype]
+                    initializers[init_name].append(init_dim)
```

Comment:
This part seems problematic to me. Is the intention to extract a list of initializers for particular node types? Currently you're checking the start of the initializer tensor name for this but that may not be reliable. A better way would be to build a list of nodes based on `op_type in ["Conv","Mul","Add"]`, then use `ModelWrapper.get_initializer()` on the inputs of each node.